libs/langchain/langchain/llms/fireworks.py

```diff
@@ -0,0 +1,372 @@
+            stop: Optional list of stop words to use when generating.
+        Returns:
+            The full LLM output.
+        """
+        params = {"model": self.model_id}
+        params = {**params, **kwargs}
+        sub_prompts = self.get_batch_prompts(params, prompts, stop)
+        choices = []
+        token_usage: Dict[str, int] = {}
+        _keys = {"completion_tokens", "prompt_tokens", "total_tokens"}
+        for _prompts in sub_prompts:
+            response = completion_with_retry(self, prompt=prompts, **params)  
+            choices.extend(response)
+            update_token_usage(_keys, response, token_usage)
+
+        return self.create_llm_result(choices, prompts, token_usage)
+
+    async def _agenerate(
+        self,
+        prompts: List[str],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> LLMResult:
+        """Call out to Fireworks endpoint async with k unique prompts."""
+        params = {"model": self.model_id}
+        params = {**params, **kwargs}
+        sub_prompts = self.get_batch_prompts(params, prompts, stop)
+        choices = []
+        token_usage: Dict[str, int] = {}
+        _keys = {"completion_tokens", "prompt_tokens", "total_tokens"}
+        for _prompts in sub_prompts:
+            response = await acompletion_with_retry(self, prompt=_prompts, **params)
+            choices.extend(response)
+            update_token_usage(_keys, response, token_usage)
+
+        return self.create_llm_result(choices, prompts, token_usage)
+
+    def get_batch_prompts(
+        self,
+        params: Dict[str, Any],
+        prompts: List[str],
+        stop: Optional[List[str]] = None,
+    ) -> List[List[str]]:
+        """Get the sub prompts for llm call."""
+        if stop is not None:
+            if "stop" in params:
+                raise ValueError("`stop` found in both the input and default params.")
+
+        sub_prompts = [
+            prompts[i : i + self.batch_size]
+            for i in range(0, len(prompts), self.batch_size)
+        ]
+        return sub_prompts
+
+    def create_llm_result(
+        self, choices: Any, prompts: List[str], token_usage: Dict[str, int]
+    ) -> LLMResult:
+        """Create the LLMResult from the choices and prompts."""
+        generations = []
+
+        for i, _ in enumerate(prompts):
+            sub_choices = choices[i : (i + 1)]
+            generations.append(
+                [
+                    Generation(
+                        text=choice,
+                    )
+                    for choice in sub_choices
+                ]
+            )
+        llm_output = {"token_usage": token_usage, "model_id": self.model_id}
+        return LLMResult(generations=generations, llm_output=llm_output)
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return "fireworks"
+
+class FireworksChat(BaseLLM):
```

Comment:
should this be implemented as a ChatModel instead of BaseLLM? (and live in langchain/chat_models)

Reply:
We implemented FireworksChat(similar to OpenAIChat). We are currently implementing a chat completion API, which we we will then use to create ChatFireworks(similar to ChatOpenAI) which will live in langchain/chat_models

Reply:
OpenAIChat is deprecated in favor of ChatOpenAI, recommend against using that as a template. if it's easier we could just take FireworksChat out for now and in sep pr add ChatFireworks (no need to block this PR)

Reply:
Sounds good to me!